src/chart/generateCategoricalChart.tsx

```diff
@@ -288,87 +288,115 @@ const getAxisMapByAxes = (
 
     let domain, duplicateDomain, categoricalDomain;
 
-    if (dataKey) {
-      // has dataKey in <Axis />
-      domain = getDomainOfDataByKey(displayedData, dataKey, type);
-
-      if (type === 'category' && isCategorical) {
-        // the field type is category data and this axis is catrgorical axis
-        const duplicate = hasDuplicate(domain);
-
-        if (allowDuplicatedCategory && duplicate) {
-          duplicateDomain = domain;
-          // When category axis has duplicated text, serial numbers are used to generate scale
-          domain = _.range(0, len);
-        } else if (!allowDuplicatedCategory) {
-          // remove duplicated category
-          domain = parseDomainOfCategoryAxis(child.props.domain, domain, child).reduce(
-            (finalDomain: any, entry: any) => (finalDomain.indexOf(entry) >= 0 ? finalDomain : [...finalDomain, entry]),
-            [],
+    /*
+     * This is a hack to short-circuit the domain creation here to enhance performance.
+     * Usually, the data is used to determine the domain, but when the user specifies
+     * a domain upfront (via props), there is no need to calculate the domain start and end,
+     * which is very expensive for a larger amount of data.
+     * The only thing that would prohibit short-circuiting is when the user doesn't allow data overflow,
+     * because the axis is supposed to ignore the specified domain that way.
+     */
+    if (type === 'number' && allowDataOverflow === true) {
+      const specifiedDomain: unknown[] | null | undefined = child.props.domain;
+      const domainStart: unknown | null | undefined = specifiedDomain?.[0];
+      const domainEnd: unknown | null | undefined = specifiedDomain?.[1];
+
+      /*
+       * The `isNumber` check is needed because the user could also provide strings like "dataMin" via the domain props.
+       * In such case, we cannot short-circuit, as we need to look at the data.
+       */
+      if (!!domainStart && !!domainEnd && isNumber(domainStart) && isNumber(domainEnd)) {
```

Comment:
Nit: It would be nice to wrap the two if statements into a single function named 

```
function isDomainComputationFromDataRequired(domain): boolean
```

That would be nice, because such a function could independently be tested and it would improve readability of the larger function.

Reply:
Done! I switched to logic to `isDomainSpecifiedByUser` though, made more sense to me.